### PR TITLE
Compile under Swift 4

### PR DESCRIPTION
**Pull request by volodg**
This applies minimal changes to allow Freddy to compile under Swift 4. It polyfills the Swift 4 overflowing `Int` add/multiply API for Swift 3.

It also tweaks some code to treat `Dictionary.map`'s tuple argument as a single argument under both Swift 3 and 4, since Swift 4 won't automatically "splat" it into two separate arguments in the transform block.

### CODE REVIEW

File: Sources/JSONParsing.swift [hunk 1]
```diff
@@ -118,9 +118,15 @@ extension JSONSerialization: JSONParserType {
     /// - parameter jsonDict: The dictionary to transform into `JSON`.
     /// - returns: An instance of `JSON` matching the dictionary.
     private static func makeJSONDictionary(_ jsonDict: [Swift.String: Any]) -> JSON {
+        #if swift(>=4.0)
```
Comment (jeremy-w) [inline Sources/JSONParsing.swift]: This shouldn't need parametrization. Even in Swift 3.1, the argument is a tuple `(key: K, value: V)`, just as in Swift 4.

If we can't get the values through argument unpacking under Swift 4, then `pair` seems a better name than `arg`.

File: Tests/FreddyTests/JSONSubscriptingTests.swift [hunk 2]
```diff
@@ -69,7 +69,11 @@ class JSONSubscriptingTests: XCTestCase {
     func testThatDictionaryOfProducesResidentsByName() {
         do {
             let residentsByName = try residentJSON.decodedDictionary(at: "residentsByName", type: Resident.self)
+            #if swift(>=4.0)
```
Comment (jeremy-w) [inline Tests/FreddyTests/JSONSubscriptingTests.swift]: Ditto here. We can use the version proposed for Swift 4 also with Swift 3.

Which is really kinda weird when you see it in practice in the REPL:

```
Welcome to Apple Swift version 3.1 (swiftlang-802.0.53 clang-802.0.42). Type :help for assistance.

  3> ["a": 1].map({ $1 })    
$R1: [Int] = 1 value {       
  [0] = 1                    
}                            
  4> ["a": 1].map({ $0.value })                            
$R2: [Int] = 1 value {       
  [0] = 1                    
} 
```

File: Tests/FreddyTests/JSONTypeTests.swift [hunk 3]
```diff
@@ -34,9 +34,15 @@ class JSONTypeTests: XCTestCase {
 
     func testCastInitializeAnyDictionary() {
         let dictionary = ["foo": 1, "bar": 2, "baz": 3]
+        #if swift(>=4.0)
+        let pairCollection = dictionary.lazy.map { arg in
```
Comment (jeremy-w) [inline Tests/FreddyTests/JSONTypeTests.swift]: Same here.

**Comment by jeremy-w** [comment]
Thanks!

The fewer version-conditional blocks we can have, the better. There already aren't a ton, but it looks like we can eliminate most of them by using the version you proposed for Swift 4 even under Swift 3, by keying into the tuple rather than relying on automatic tuple unpacking.

The overflow-related code is a significant API shift. I'd prefer to reduce the amount of duplication across version branches for what's really just a method call, but I'd want to see what sort of impact, say, shimming Swift 3 to look like Swift 4, or indirecting through a multiply-and-gripe API we control, might have on performance.

**Comment by jeremy-w** [comment]
I'm updating this to track the latest `master`, and then it should be good to merge once Travis gives the A-OK. The review comments can be addressed later.

Thanks again! :)

**Comment by jeremy-w** [comment]
@volodg Getting compiler errors. Maybe the new IntExtensions shim file didn't get added to the targets?

```
❌  /Users/travis/build/bignerdranch/Freddy/Sources/JSONParser.swift:459:56: value of type 'Int' has no member 'multipliedReportingOverflow'

guard case let (exponent, .none) = 10.multipliedReportingOverflow(by: value) else {

                                                       ^~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~

❌  /Users/travis/build/bignerdranch/Freddy/Sources/JSONParser.swift:483:52: value of type 'Int' has no member 'multipliedReportingOverflow'

guard case let (signedValue, .none) = sign.rawValue.multipliedReportingOverflow(by: value) else {

                                              ~~~~~^~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by volodg** [comment]
@jeremy-w you are right, fixed

**Pull request merged**